### PR TITLE
[WIP] Add support for FixedPrecisionTensor chain [Replicated Sharing Tensor (RST)]

### LINKED ISSUE #3931: Replicated Sharing Tensor (RST): Add support for FixedPrecisionTensor chain
Replicated Sharing Tensor currently can work if the plain text type is float or int, we want it to be able to work given a FixedPrecisionTensor as plain text.
When this issue is solved this test case should work properly:
```
a = a.fix_perc().share(bob, alice, james, protocol = "falcon")
b = a.reconstruct().float_perc()
assert (a==b).all()
```
also, it should be tested on arithmetic operation (addition, multiplication, and matrix multiplication)

### PR DESCRIPTION
**Pull request by aanurraj**
## Description
Added support for the FPT chain in RST and updated add and sub to support these operations in FPT chain. 

On completion this PR closes #3931 

## Affected Dependencies
None

## How has this been tested?
for [Replicated Sharing Tensor (RST)] FixedPrecisionTensor chain 
> Test cases are added for .share while using precision
> Test cases are added for add and sub  

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: syft/frameworks/torch/tensors/interpreters/native.py [hunk 1]
```diff
@@ -929,17 +924,26 @@ def share(
...
-                    field=field,
-                    dtype=dtype,
-                    crypto_provider=crypto_provider,
-                    owner=self.owner,
-                )
-                .on(self.copy(), wrap=False)
-                .share_secret(*owners)
-            )
+            else:
+                if self.type() == "torch.FloatTensor":
```
Comment (abogaziah) [inline syft/frameworks/torch/tensors/interpreters/native.py:928]: this condition is not reachable, you have already checked it previously.

File: syft/frameworks/torch/tensors/interpreters/native.py [hunk 2]
```diff
@@ -1103,3 +1107,5 @@ def numpy_tensor(self):
                 "on a wrapper. Add NumpyTensor to the chain by hand if you want "
                 "this functionality.",
             )
+
+    reconstruct = get
```
Comment (abogaziah) [inline syft/frameworks/torch/tensors/interpreters/native.py:1111]: how is that useful?

File: test/torch/tensors/test_replicated_shared.py [hunk 3]
```diff
@@ -180,3 +180,142 @@ def test_workers_arrangement(workers):
     me, bob, alice = (workers["me"], workers["bob"], workers["alice"])
     x = torch.tensor(7).share(bob, alice, me, protocol="falcon")
     assert x.players[0] == me
+
+
+def test_fixed_precision_and_sharing(workers):
```
Comment (abogaziah) [inline test/torch/tensors/test_replicated_shared.py:185]: fixed pointed arithmetic is not test thought it's the purpose of that issue, numbers after decimal point should

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py [hunk 4]
```diff
@@ -306,3 +309,7 @@ def __str__(self):
             for v in self.child.values():
                 out += "\n\t-> " + str(v)
         return out
+
+
+### Register the tensor with hook_args.py ###
```
Comment (abogaziah) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:314]: what is that?

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py [hunk 5]
```diff
@@ -67,6 +68,8 @@ def reconstruct(self):
         plain_text = self.__map_modular_to_real(plain_text_mod)
         return plain_text
 
+    get = reconstruct
```
Comment (abogaziah) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:71]: how is that useful?

File: syft/frameworks/torch/tensors/interpreters/precision.py [hunk 6]
```diff
@@ -194,6 +194,9 @@ def add(self, _self, other):
             # we swap operators so that we do the same operation as above
             _self, other = other, _self.wrap()
 
+        elif isinstance(other, ReplicatedSharingTensor) and isinstance(_self, torch.Tensor):
```
Comment (abogaziah) [inline syft/frameworks/torch/tensors/interpreters/precision.py:197]: this is tight coupling which should be avoided, but why would you swap them anyway?

File: syft/frameworks/torch/tensors/interpreters/precision.py [hunk 7]
```diff
@@ -234,6 +237,9 @@ def sub(self, _self, other):
             # we swap operators so that we do the same operation as above
             _self, other = -other, -_self.wrap()
 
+        elif isinstance(other, ReplicatedSharingTensor) and isinstance(_self, torch.Tensor):
```
Comment (abogaziah) [inline syft/frameworks/torch/tensors/interpreters/precision.py:240]: why would you mul -1?

File: syft/frameworks/torch/tensors/interpreters/native.py (hunk 2 above)
Comment (aanurraj) [reply syft/frameworks/torch/tensors/interpreters/native.py:1111]: this was added temporarily as few of the tests are written using reconstruct(), which fails as it gets called on native.py, in future, it will be good to go with .get() instead of reconstruct()

File: syft/frameworks/torch/tensors/interpreters/native.py (hunk 1 above)
Comment (aanurraj) [reply syft/frameworks/torch/tensors/interpreters/native.py]: Sorry, fixed in latest commit

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 4 above)
Comment (aanurraj) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:314]: its been added so that the  _hook_method_args of overload_method can know Put back SyftTensor on the tensors found in the response. if this was not used the output of add or sub would have (Wrapper)>[ReplicatedSharingTensor] instead of (Wrapper)>FixedPrecisionTensor>[RaplicatedSharingTensor] after the operation.

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 5 above)
Comment (aanurraj) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:71]: a.reconstruct() <-- this is a wrapper
and in native we do not have the reconstruct method
so it fails, thus as I mentioned earlier we should use .get() universally in future.

File: syft/frameworks/torch/tensors/interpreters/precision.py (hunk 7 above)
Comment (aanurraj) [reply syft/frameworks/torch/tensors/interpreters/precision.py:240]: because putting "-" was giving an error (i can't remember the error, for now, I'll tell you soon though.

File: syft/frameworks/torch/tensors/interpreters/precision.py (hunk 6 above)
Comment (aanurraj) [reply syft/frameworks/torch/tensors/interpreters/precision.py:197]: the reason can be found in the above code for AST, the swap has been done for the same reason. 
now about

File: syft/frameworks/torch/tensors/interpreters/native.py (hunk 2 above)
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/native.py:1108]: I disagree, you "get" a remote tensor but you "reconstruct" a shared tensor, calling everything get is the easy way, not the correct one, 
can you explain further? give an example of a test case that fails?

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 4 above)
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:314]: this is not the right place, it breaks the level of abstraction torch hook is a lower-level abstraction so RST should not know it exists.

File: syft/frameworks/torch/tensors/interpreters/native.py (hunk 2 above)
Comment (aanurraj) [reply syft/frameworks/torch/tensors/interpreters/native.py:1108]: Screenshot 2020-08-30 at 12 17 27 AM

if that is not included  this error will prompt up with "reconstruct"

File: test/torch/tensors/test_replicated_shared.py (hunk 3 above)
Comment (aanurraj) [reply test/torch/tensors/test_replicated_shared.py:185]: fixed

File: syft/frameworks/torch/tensors/interpreters/precision.py (hunk 7 above)
Comment (aanurraj) [reply syft/frameworks/torch/tensors/interpreters/precision.py:240]: Screenshot 2020-08-30 at 12 31 19 AM
here is the error!

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 4 above)
Comment (aanurraj) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:314]: @gmuraru can you help me this issue?

Comment (gmuraru) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:314]: Yep - this is needed because when we add the wrapper we check what tensors exists - if we do not have this here it would not know how to create the ```(Wrapper)>RST```

Comment (gmuraru) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:314]: A better approach to this (for all syft framework) it should be to add something "invasive" like:
```
1. get all classes that inherit from AbstractTensor
2. call this ```default_register_tensor``` on all of them
```
But I think another issue might be in order to do this -- doing one now.

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 5 above)
Comment (gmuraru) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:71]: Hmm...thinking of this. Could we add a reconstruct method in "native" and have a check -- we can do the reconstruct only if we have ```AdditiveSharingTensor``` or ```ReplicatedSharingTensor```

File: syft/frameworks/torch/tensors/interpreters/native.py (hunk 2 above)
Comment (aanurraj) [reply syft/frameworks/torch/tensors/interpreters/native.py]: fixed in the latest commit!

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 5 above)
Comment (aanurraj) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py]: fixed in latest commit!

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 4 above)
Comment (aanurraj) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:312]: we are doing this in other issue, right? @gmuraru @abogaziah

File: syft/frameworks/torch/tensors/interpreters/native.py [hunk 8]
```diff
@@ -929,17 +929,23 @@ def share(
...
-                    protocol=protocol,
-                    field=field,
-                    dtype=dtype,
-                    crypto_provider=crypto_provider,
-                    owner=self.owner,
-                )
-                .on(self.copy(), wrap=False)
-                .share_secret(*owners)
-            )
+            else:
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/native.py:932]: I think you can remove the else here - if it enters the throw

Comment (aanurraj) [reply syft/frameworks/torch/tensors/interpreters/native.py:939]: done :)

**Comment by abogaziah** [comment]
update black and run it on your fork, or sync it with upstream/master, changed files currently 69, it should be around 5 when you do that successfully
